src/lib/storage/disk_analyzer.rb

```diff
@@ -0,0 +1,368 @@
+#!/usr/bin/env ruby
+#
+# encoding: utf-8
+
+# Copyright (c) [2015] SUSE LLC
+#
+# All Rights Reserved.
+#
+# This program is free software; you can redistribute it and/or modify it
+# under the terms of version 2 of the GNU General Public License as published
+# by the Free Software Foundation.
+#
+# This program is distributed in the hope that it will be useful, but WITHOUT
+# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+# FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for
+# more details.
+#
+# You should have received a copy of the GNU General Public License along
+# with this program; if not, contact SUSE LLC.
+#
+# To contact SUSE LLC about this file by physical or electronic mail, you may
+# find current contact information at www.suse.com.
+
+require "yast"
+require "fileutils"
+require "pp"
+
+module Yast
+  module Storage
+    #
+    # Class to analyze the disks (the storage setup) of the existing system:
+    # Check the existing disks and their partitions what candidates there are
+    # to install on, typically eliminate the installation media from that list
+    # (unless there is no other disk), check if there already are any
+    # partitions that look like there was a Linux system previously installed
+    # on that machine, check if there is a Windows partition that could be
+    # resized.
+    #
+    # Some of those operations involve trying to mount the underlying
+    # filesystem.
+    #
+    class DiskAnalyzer
+      include Yast::Logger
+
+      LINUX_PARTITION_IDS =
+        [
+          ::Storage::ID_LINUX,
+          ::Storage::ID_SWAP,
+          ::Storage::ID_LVM,
+          ::Storage::ID_RAID
+        ]
+
+      WINDOWS_PARTITION_IDS =
+        [
+          ::Storage::ID_NTFS,
+          ::Storage::ID_DOS32,
+          ::Storage::ID_DOS16,
+          ::Storage::ID_DOS12
+        ]
+
+      DEFAULT_DISK_CHECK_LIMIT = 10
+
+      attr_reader :installation_disks, :candidate_disks
+      attr_reader :windows_partitions, :linux_partitions
+      attr_accessor :disk_check_limit
+
+      def initialize
+        Yast.import "Arch"
+
+        @installation_disks = []
+        @candidate_disks    = []
+        @linux_partitions   = []
+        @windows_partitions = [] # only filled if @linux_partitions is empty
+
+        # Maximum number of disks to check. This might be important on
+        # architectures that tend to have very many disks (s390).
+        @disk_check_limit = DEFAULT_DISK_CHECK_LIMIT
+      end
+
+      # Analyze disks and partitions. Make sure to call this before querying
+      # any member variables.
```

Comment:
well, it is not 100% you already sneak in logic to not analyze windows if there is linux partitions, which is wrong from my POV as it should not be decision of this class. Be in control also means, that you can have potential problems like below if you forgot to pass already analyzed object, it will return empty arrays, which looks even more wrong for me, as it is not immediatell obvious what is problem.
What is reason to have control over mounting? for me it is nature to get exception if I call windows_partitions if something wrong happen when trying to detect it, so I know that information is unreliable from that method ( as opposed to current situation when it can be empty if I forgot analyze or only partial if some mount failed ).